File: app/src/main/java/com/example/mealrecipeapp/ui/home/HomeViewModel.java
package com.example.mealrecipeapp.ui.home;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.mealrecipeapp.data.remote.response.Recipe;
import com.example.mealrecipeapp.data.repository.AppRepository;
import com.example.mealrecipeapp.utils.Resource;

import java.util.List;
import java.util.Objects;

public class HomeViewModel extends ViewModel {
    private final AppRepository appRepository;
    private final MutableLiveData<Resource<List<Recipe>>> recipes = new MutableLiveData<>();
    private final MutableLiveData<String> query = new MutableLiveData<>("");

    public HomeViewModel(AppRepository appRepository) {
        this.appRepository = appRepository;
        searchRecipes("");
    }

    public void setQuery(String newQuery) {
        if (query.getValue() != null &&!query.getValue().equals(newQuery) && !newQuery.equals("")) {
            query.postValue(newQuery);
            searchRecipes(newQuery);
        }
    }
    private void searchRecipes(String query) {
        appRepository.getRecipes(query).observeForever(recipes::postValue);
    }

    public LiveData<Resource<List<Recipe>>> getRecipesLiveData() {
        return recipes;
    }

    public String getUserName() {
        return appRepository.getUserName();
    }

    public String getUserImage() {
        return appRepository.getUserImage();
    }
}

File: app/src/main/java/com/example/mealrecipeapp/ui/home/HomeViewModel.kt
package com.example.mealrecipeapp.ui.home

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.mealrecipeapp.data.remote.response.Recipe
import com.example.mealrecipeapp.data.repository.AppRepository
import com.example.mealrecipeapp.utils.Resource
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class HomeViewModel(private val appRepository: AppRepository) : ViewModel() {
    private val _recipes = MutableLiveData<Resource<List<Recipe>>>()
    val recipes: LiveData<Resource<List<Recipe>>>
        get() = _recipes

    private val query = MutableLiveData("")

    private var currentJob: Job? = null

    init {
        searchRecipes("")
    }

    fun setQuery(newQuery: String) {
        if (query.value != newQuery) {
            query.postValue(newQuery)
            searchRecipes(newQuery)
        }
    }

    private fun searchRecipes(query: String) {
        currentJob?.cancel()
        currentJob = viewModelScope.launch {
            delay(800L)
            _recipes.postValue(Resource.Loading())
            try {
                val listRecipes = appRepository.getRecipes(query)
                _recipes.postValue(Resource.Success(listRecipes))
            } catch (e: Exception) {
                _recipes.postValue(Resource.Error(e.message.toString()))
            }
        }
    }
    fun getUserName(): String = appRepository.getUserName()

    fun getUserImage(): String = appRepository.getUserImage()

    override fun onCleared() {
        super.onCleared()
        currentJob?.cancel()
    }
}
